### ISSUE TITLE
statistics: choose wrong index when no correlate column exist

### ISSUE THREAD

**Issue by L-maple** [pingcap/tidb#45915]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```SQL
CREATE TABLE `push_record` (
  `poi_id` bigint(20) NOT NULL,
  `biz_name` varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin NOT NULL,
  `template_id` varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin NOT NULL,
  `template_data_belong_date` varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin NOT NULL,
  `send_status` int(11) NOT NULL,
  `is_del` int(11) NOT NULL,
  `add_time` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
  `update_time` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  KEY `idx_poi_id` (`poi_id`),
  KEY `idx_update_time` (`update_time`),
  KEY `idx_biz_name_template_id_template_data_belong_date` (`biz_name`,`template_id`,`template_data_belong_date`),
  UNIQUE KEY `uniq_poi_id_biz_name_template_id_template_data_belong_date` (`poi_id`,`biz_name`,`template_id`,`template_data_belong_date`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci/*!90000 SHARD_ROW_ID_BITS=4 */;
```
```SQL
select  poi_id,biz_name,template_id,template_data_belong_date,send_status,is_del,add_time,update_time  from push_record 
  where biz_name = 'bizReport' and is_del= 0  and template_id = '100181004'  and template_data_belong_date = '2023-05-15'  and send_status in (3) 
  order by poi_id asc limit 0, 1000;
```

### 2. What did you expect to see? (Required)
I dump the statistics, but I cannot paste it here.
But the problem is, for this query, `idx_biz_name_template_id_template_data_belong_date` is better than `uniq_poi_id_biz_name_template_id_template_data_belong_date`.

### 3. What did you see instead (Required)
But the optimizer chooses `uniq_poi_id_biz_name_template_id_template_data_belong_date`. the index range is [NULL, +inf)

### 4. What is your TiDB version? (Required)
master

**Comment by aytrack** [comment]
i can't reproduce on master
```sql
[10:52:50]TiDB root:test> explain select  poi_id,biz_name,template_id,template_data_belong_date,send_status,is_del,add_time,update_time  from push_record
                       ->   where biz_name = 'bizReport' and is_del= 0  and template_id = '100181004'  and template_data_belong_date = '2023-05-15'  and send_status in (3)
                       ->   order by poi_id asc limit 0, 1000;
+------------------------------+---------+-----------+-------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------+
| id                           | estRows | task      | access object                                                                                                                 | operator info                                                                                                     |
+------------------------------+---------+-----------+-------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------+
| TopN_10                      | 0.00    | root      |                                                                                                                               | test.push_record.poi_id, offset:0, count:1000                                                                     |
| └─IndexLookUp_30             | 0.00    | root      |                                                                                                                               |                                                                                                                   |
|   ├─IndexRangeScan_27(Build) | 0.00    | cop[tikv] | table:push_record, index:idx_biz_name_template_id_template_data_belong_date(biz_name, template_id, template_data_belong_date) | range:["bizReport" "100181004" "2023-05-15","bizReport" "100181004" "2023-05-15"], keep order:false, stats:pseudo |
|   └─Selection_29(Probe)      | 0.00    | cop[tikv] |                                                                                                                               | eq(test.push_record.is_del, 0), eq(test.push_record.send_status, 3)                                               |
|     └─TableRowIDScan_28      | 0.00    | cop[tikv] | table:push_record                                                                                                             | keep order:false, stats:pseudo                                                                                    |
+------------------------------+---------+-----------+-------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------+
5 rows in set
Time: 0.012s
[10:53:17]TiDB root:test> select tidb_version();
+-----------------------------------------------------------+
| tidb_version()                                            |
+-----------------------------------------------------------+
| Release Version: v7.4.0-alpha                             |
| Edition: Community                                        |
| Git Commit Hash: cdeac4c66f728ad60a5166ef7700efa52cbd1384 |
| Git Branch: heads/refs/tags/v7.4.0-alpha                  |
| UTC Build Time: 2023-08-08 14:24:50                       |
| GoVersion: go1.20.7                                       |
| Race Enabled: false                                       |
| Check Table Before Drop: false                            |
| Store: tikv                                               |
+-----------------------------------------------------------+
1 row in set
Time: 0.003s
```

**Comment by L-maple** [comment]
@aytrack 
case 1: set tidb_opt_prefer_range_scan=0
![image](https://github.com/pingcap/tidb/assets/20724552/80a8fb88-1e66-40e8-ae96-70d25ab23d69)
TableScan is terrible.
case 2: set tidb_opt_prefer_range_scan=1
![image](https://github.com/pingcap/tidb/assets/20724552/c74da841-1db0-4b09-8605-db57015c1988)
The plan is good just because the bad index is removed, but bad index has lower cost. 

My PR is to solve WHY bad index should has lower cost:
bad_index cost:  1068481061.13
![image](https://github.com/pingcap/tidb/assets/20724552/b0d85991-67c5-4416-b8ee-93ffbb9223c3)
good_index_cost:  4153587992.22
![image](https://github.com/pingcap/tidb/assets/20724552/c1392857-9175-4320-88ed-24210e816ac0)

**Comment by time-and-fate** [comment]
Sorry for being late, and thanks for your investigation and contribution.

Since we are unable to reproduce your problem, I have to speculate from the information you provide.

#### About this problem
The main problem here is the choice between the index that satisfies the `WHERE` filter and the index that satisfies the `ORDER BY` columns.

To make it more clear, let's see an example:
```sql
CREATE TABLE `t` (
  `a` int DEFAULT NULL,
  `b` int DEFAULT NULL,
  KEY `ia` (`a`),
  KEY `ib` (`b`)
);
select * from t where b > 900000 order by a limit 1;
```
```
> explain select * from t1 use index (ib) where b > 900000 order by a limit 1;
+----------------------------------+-----------+-----------+-----------------------+---------------------------------------+
| id                               | estRows   | task      | access object         | operator info                         |
+----------------------------------+-----------+-----------+-----------------------+---------------------------------------+
| TopN_9                           | 1.00      | root      |                       | test.t1.a, offset:0, count:1          |
| └─IndexLookUp_16                 | 1.00      | root      |                       |                                       |
|   ├─IndexRangeScan_13(Build)     | 100284.60 | cop[tikv] | table:t1, index:ib(b) | range:(900000,+inf], keep order:false |
|   └─TopN_15(Probe)               | 1.00      | cop[tikv] |                       | test.t1.a, offset:0, count:1          |
|     └─TableRowIDScan_14          | 100284.60 | cop[tikv] | table:t1              | keep order:false                      |
+----------------------------------+-----------+-----------+-----------------------+---------------------------------------+

> explain select * from t1 use index (ia) where b > 900000 order by a limit 1;
+-----------------------------------+---------+-----------+-----------------------+-----------------------+
| id                                | estRows | task      | access object         | operator info         |
+-----------------------------------+---------+-----------+-----------------------+-----------------------+
| Limit_12                          | 1.00    | root      |                       | offset:0, count:1     |
| └─Projection_22                   | 1.00    | root      |                       | test.t1.a, test.t1.b  |
|   └─IndexLookUp_21                | 1.00    | root      |                       |                       |
|     ├─IndexFullScan_18(Build)     | 9.97    | cop[tikv] | table:t1, index:ia(a) | keep order:true       |
|     └─Selection_20(Probe)         | 1.00    | cop[tikv] |                       | gt(test.t1.b, 900000) |
|       └─TableRowIDScan_19         | 9.97    | cop[tikv] | table:t1              | keep order:false      |
+-----------------------------------+---------+-----------+-----------------------+-----------------------+
```

It's obvious that the execution time for using ib is mostly decided by the selectivity of the condition.
But for index(a), it's more complicated. If the first n rows we read just all satisfy the condition, we can stop reading immediately. But if the data we need is at the end of ia, we need to use IndexLookup to read almost the entire table.

So the core problem is the estimation for the row count of the index scan on index ia. Unfortunately, it's a very hard problem. Based on currently available statistics, it's almost impossible to guarantee a precise estimation in all scenarios.

#### About the code here
You correctly located the place where this estimation logic is handled. But the bad estimation here is not because of the use of correlation. Actually, the correlation doesn't take effect here. 
When `corr` is 0 here, the estimation formula becomes `prop.ExpectedCnt/selectivity`, which is tidb's current estimation strategy for the `Limit -> Selection -> Scan` structure where the expected row count of `Limit` is known.
As we said above, currently, no estimation strategy can fit all scenarios here.

#### Improvements in TiDB
We have been aware of this problem. So from v7.0.0, we provided a system variable to control this behavior.

You can refer to the user doc: [English](https://docs.pingcap.com/tidb/dev/system-variables#tidb_opt_ordering_index_selectivity_threshold-new-in-v700)  [Chinese](https://docs.pingcap.com/zh/tidb/dev/system-variables#tidb_opt_ordering_index_selectivity_threshold-%E4%BB%8E-v700-%E7%89%88%E6%9C%AC%E5%BC%80%E5%A7%8B%E5%BC%95%E5%85%A5)
In short, you run this SQL:
```sql
SET GLOBAL tidb_opt_ordering_index_selectivity_threshold = 1;
```
This should make tidb behaves the same as in your PR.
You can give it a try and see if this fix your problem.

**Comment by L-maple** [comment]
@time-and-fate Thanks for your analysis! 
`tidb_opt_ordering_index_selectivity_threshold` is quite useful. I will close the PR and this issue.
